sqlx-core/src/postgres/types/chrono/datetime.rs

```diff
@@ -101,21 +104,52 @@ impl<Tz: TimeZone> Encode<'_, Postgres> for DateTime<Tz> {
 
 impl<'r> Decode<'r, Postgres> for DateTime<Local> {
     fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
-        let naive = <NaiveDateTime as Decode<Postgres>>::decode(value)?;
-        Ok(Local.from_utc_datetime(&naive))
+        Ok(<DateTime<Utc> as Decode<Postgres>>::decode(value)?.with_timezone(&Local))
     }
 }
 
 impl<'r> Decode<'r, Postgres> for DateTime<Utc> {
     fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
-        let naive = <NaiveDateTime as Decode<Postgres>>::decode(value)?;
-        Ok(Utc.from_utc_datetime(&naive))
+        Ok(match value.format() {
+            PgValueFormat::Binary => {
+                // TIMESTAMP is encoded as the microseconds since the epoch
+                let epoch = Utc.ymd(2000, 1, 1).and_hms(0, 0, 0);
+                let us = Decode::<Postgres>::decode(value)?;
+                epoch + Duration::microseconds(us)
+            }
+
+            PgValueFormat::Text => {
+                let s = value.as_str()?;
+                DateTime::parse_from_str(
+                    s,
+                    if is_timestamptz_output(s) {
+                        // Note that because the output format differs from RFC3339,
+                        // `DateTime::parse_from_rfc3339` is not available
+                        "%Y-%m-%d %H:%M:%S%.f%#z"
+                    } else {
+                        "%Y-%m-%d %H:%M:%S%.f"
+                    },
+                )?
+                .with_timezone(&Utc)
+            }
+        })
     }
 }
 
 impl<'r> Decode<'r, Postgres> for DateTime<FixedOffset> {
     fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
-        let naive = <NaiveDateTime as Decode<Postgres>>::decode(value)?;
-        Ok(Utc.fix().from_utc_datetime(&naive))
+        let datetime_naive_utc = <DateTime<Utc> as Decode<Postgres>>::decode(value)?.naive_utc();
+        Ok(Utc.fix().from_utc_datetime(&datetime_naive_utc))
```

Comment:
This isn't doing what I think you intended it to do. In #703 you said:

> For `DateTime<FixedOffset>`, the existing implementation implies that the timezone of the output follows the local time of the server where the application is running.

This is half-true. The binary encoding gives the timestamp at `+00:00` (UTC) offset; we can't know what the server's timezone is unless we look it up.

However, the text encoding _does_ give the offset, which `DateTime<FixedOffset>` can represent. However, because you've implemented this on top of `DateTime<Utc>`, it will always be in UTC.

What I would recommend is moving the code for `DateTime<Utc>` (which does the actual parsing) to here, and then in the `DateTime<Utc>` impl, doing:

```rust
Ok(DateTime::<FixedOffset>::decode(value)?.with_timezone(&Utc))
```